google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/Publisher.java

```diff
@@ -209,6 +240,10 @@ public String getTopicNameString() {
     messagesWaiter.incrementPendingMessages(1);
 
     if (!batchesToSend.isEmpty()) {
+      // TODO: if this is not an ordering keys scenario, will this do anything?
+      publishAllWithoutInflight();
+
+      // TODO: if this is an ordering keys scenario, is this safe without messagesBatchLock?
```

Comment:
> is this safe without messagesBatchLock?
Since we are assuming publish() for the same ordering key is not called by multiple threads, I think it would be safe without the lock. If users use multiple threads for publishing an ordering key, we cannot guarantee the order anyway. However, it is always possible that I am missing something. WDYT? (But, the executor below should still be removed. Please see my another comment below)

Reply:
I don't think we are assuming this. People could call publish for the same ordering key from multiple threads. It just means that the order among those two messages can not be known by the publisher. It will still be established (and will in fact be established in the order in which they are sent from the publisher client to Cloud Pub/Sub.

Reply:
Right, users can definitely use threads for the same ordering key if they want, and we should support it. I think what I wanted to say was that if a user calls publish() for the same ordering key concurrently using threads without a proper synchronization, then there is no ordering in that behavior anyway.
I actually looked into the code again, and I think I was wrong. We should get messagesBatchLock. Here is the scenario:
(1) By messagesBatch.add(outstandingPublish) at line #231, `outstandingPublish` is added to `messages`, and an old batch is returned.
(2) Before the returned batch is published, publishAllWithoutInflight() is called by the timer.
(3) publishAllWithoutInflight() publishes the message added to the current batch at (1). Now, the new message is published before the old batch is published.